metafix/src/main/java/org/metafacture/metafix/FixMethod.java

```diff
@@ -84,6 +85,24 @@ public void apply(final Metafix metafix, final Record record, final List<String>
             metafix.putMap(params.get(0), options);
         }
     },
+    put_rdfmap {
+        @Override
+        public void apply(final Metafix metafix, final Record record, final List<String> params, final Map<String, String> options) {
+            final String rdfMapName = params.size() == 1 ? params.get(0) : params.get(1) + options.get(RdfMap.TARGET) + options.getOrDefault(RdfMap.TARGET_LANGUAGE, "");
```

Comment:
Conceptually a triple is reduced to key-value. The reduction depends on `target[_language]`. I.e. there could be different maps coming from the same source but with different  `target[_language]`. So internally the name of the map reflects that by appending the `target[_language]` to its name.

Reply:
But you're only changing the supplied map name, if present, not the file name.

I really think we should just let the user deal with it, at least for now.

Reply:
Using the same file name several different maps can be constructed. That is why this is needed.

Reply:
Only if the user doesn't give them unique names. It's essentially the same with file maps where we don't take different parameters into account either.

But if you feel that we should handle it transparently, I'll try to come with a different approach.

Reply:
In case we do want transparent option handling, I suggest that _implicit_ names are derived from the file/resource name and some (all?) options, but _explicit_ names are left untouched. The same should apply to _all_ external maps.

Reply:
Ah, I think I understand what you mean - it was like this before I realized that `put_rdfmap` would not be coherent to `lookup_rdf`. So I would revert that. I like that very much, because it becomes to what made me exclaim "elegant" :)

Reply:
I'm sorry, your answer is not entirely clear to me:

1. Are you going to update your implementation?
2. Do you think that we should include options in implicitly named maps?

Reply:
1. yes
2. yes, `target` is mandatory

Reply:
Ohm wait ... 2) we may not need at all. Will think about it Monday.

Reply:
Okay. Regarding 2) I'd like to get feedback from @fsteeg and @TobiasNx as well: Do you agree that we should change the behaviour for _implicitly_ named file maps? If so, I'll implement it after this pull request has been merged.

> In case we do want transparent option handling, I suggest that _implicit_ names are derived from the file/resource name and some (all?) options, but _explicit_ names are left untouched. The same should apply to _all_ external maps.

Reply:
Another thought that crossed my mind: We're going to some lengths to implement `lookup_rdf()`, but do we really need it? Are we going to need specialized lookup methods for every type of map we might want to introduce?[^1]

Why not just `put_rdfmap()` and then `lookup()` as usual? No special treatment necessary.[^2] This might even make it clearer that you need unique names for differently configured maps (except when you actually want to overwrite an existing map).[^3]

```perl
put_rdfmap("path.rdf", target: "foo")
put_rdfmap("path.rdf", target: "bar") # oops!
lookup("field", "path.rdf")

put_rdfmap("path.rdf", "foo-map", target: "foo")
put_rdfmap("path.rdf", "bar-map", target: "bar") # better!
lookup("field", "foo-map")
```

[^1]: I'm currently in the process of introducing LMDB maps in Limetrans, but it never even occurred to me that I might need a `lookup_lmdb()` method. I just add the map (from Java in this case) and be done with it.
[^2]: `lookup()` only has built-in support for file maps due to compatibility with Catmandu, which doesn't have a concept of separately defined maps.
[^3]: I might even go so far as to suggest making an explicit name _mandatory_ for `put_*map()`.

Reply:
> Regarding 2) I'd like to get feedback from @fsteeg and @TobiasNx as well [...]

> I suggest that implicit names are derived from the file/resource name and some (all?) options, but explicit names are left untouched. The same should apply to all external maps.

Sounds good, +1.

> Why not just `put_rdfmap()` and then `lookup()` as usual?

Yes, and as I wrote [here](https://github.com/metafacture/metafacture-core/issues/415#issuecomment-1145907752) & [here](https://github.com/metafacture/metafacture-core/issues/415#issuecomment-1148270176), I think something like `lookup(field, 'rpb.ttl')` should work even without a previous `put_rdfmap()` (i.e. both `target` and `targetLanguage` being optional, defaulting `target` to `prefLabel` and maybe erroring if there are multiple languages but no `targetLanguage`).

(btw `targetLanguage` is quite verbose, wouldn't `language` or even `lang` be good?)

Reply:
> Sounds good, +1.

Would you still support this implicit naming if we dropped `lookup_rdf()`? I feel it would be much simpler if all maps required explicit names. The _only_ exception would be `lookup("field", "path.csv")` due to its compatibility requirement.

> I think something like `lookup(field, 'rpb.ttl')` should work even without a previous `put_rdfmap()`

Hm, that's actually contrary to my proposal. Do you mean we should somehow detect which `put_*map()` to call based on the path name?

> i.e. both `target` and `targetLanguage` being optional

This is orthogonal to the issue at hand. Both _can_ be optional for `put_rdfmap()`, but `lookup()` shouldn't take any options (except for local/unnamed maps).

Reply:
> > Regarding 2) I'd like to get feedback from @fsteeg and @TobiasNx as well [...]
> 
> > I suggest that implicit names are derived from the file/resource name and some (all?) options, but explicit names are left untouched. The same should apply to all external maps.
> 
> Sounds good, +1.
In my opinion +1 too

> 
> > Why not just `put_rdfmap()` and then `lookup()` as usual?
> 
> Yes, and as I wrote [here](https://github.com/metafacture/metafacture-core/issues/415#issuecomment-1145907752) & [here](https://github.com/metafacture/metafacture-core/issues/415#issuecomment-1148270176), I think something like `lookup(field, 'rpb.ttl')` should work even without a previous `put_rdfmap()` (i.e. both `target` and `targetLanguage` being optional, defaulting `target` to `prefLabel` and maybe erroring if there are multiple languages but no `targetLanguage`).

There are alternative `lookup`s in Catmandu `looup_in_store`: https://metacpan.org/pod/Catmandu::Fix::lookup_in_store. I voted for a specific skos_lookup since you can use multiple queries with the same file: https://github.com/metafacture/metafacture-core/issues/415#issuecomment-1145924854
If we go with a put_rdfmap we should recreate the scenarios introduced in the ticket: https://github.com/metafacture/metafacture-core/issues/415#issue-1043693019

> 
> (btw `targetLanguage` is quite verbose, wouldn't `language` or even `lang` be good?)
The reason for that ist that also the source element _can_ have a language that is targeted with `matchLanguage`. See: https://github.com/metafacture/metafacture-core/issues/415#issue-1043693019

Reply:
> I voted for a specific skos_lookup since you can use multiple queries with the same file

You still can. Just define multiple maps:

```perl
put_rdfmap("https://raw.githubusercontent.com/dini-ag-kim/hochschulfaechersystematik/master/hochschulfaechersystematik.ttl", "sys-de", target: "prefLabel", targetLanguage: "de")
put_rdfmap("https://raw.githubusercontent.com/dini-ag-kim/hochschulfaechersystematik/master/hochschulfaechersystematik.ttl", "sys-en", target: "prefLabel", targetLanguage: "en")
lookup("learningResourceType[].*.prefLabel.de", "sys-de")
lookup("learningResourceType[].*.prefLabel.en", "sys-en")
```

> If we go with a put_rdfmap we should recreate the scenarios introduced in the ticket and the options requested

`lookup_rdf()` is just `put_rdfmap()` + `lookup()` under the hood, they're functionally the same.

Reply:
> I feel it would be much simpler if all maps required explicit names. The _only_ exception would be `lookup("field", "path.csv")` due to its compatibility requirement. [...] [EDIT: Spoke too soon... Options are required for file maps as well; again, compatibility.]

From a user perspective my personal preference would be to have simple lookups without first declaring a named map, or configuring details for all these cases: `lookup("field", "path.csv")`, `lookup("field", "path.tsv")`, `lookup("field", "path.ttl")`, `lookup("field", de: Deutsch, en: Englisch)`. If that gets messy on the implementation side I'm not opposed to a different approach.

> Do you mean we should somehow detect which put_*map() to call based on the path name?

Yes, I was thinking that was already happening for `csv`/`tsv`, but that actually requires the `sep_char` option. I think `.csv` should default to `sep_char=','` and `.tsv` to `sep_char='\t'`. And `.ttl` should default to `put_rdfmap` with `target='prefLabel'`.

Reply:
Okay, at least we seem to be clear on the alternatives then :)

I'm not in favour of basing any behaviour on file extensions (we already have a similar situation with Fix file vs. Fix string), but I won't oppose the decision if you all agree. I just won't use it myself ;)

The next step here would probably be to integrate `lookup_rdf()` into `lookup()` by delegating to `put_rdfmap()` if the file extension is `.ttl` and to `put_filemap()` otherwise. Any implicit map names should be derived from the file name and a canonical representation of the given options; explicit map names are left untouched.

Reply:
> The next step here would probably be to integrate lookup_rdf() into lookup() by delegating to put_rdfmap() if the file extension is .ttl and to put_filemap() otherwise.

_If_ you want that really, I would rather do it the other way around, because there are many more RDF extensions than CSV derivates I guess (which are all supported by jena model (really all? Which ones exactly?). What if you have a file without any extension? Or what if you also want to support a csv accessible as a HTTP resource (I know this wouldn't be put_filemap())?

Reply:
Incorporated `lookup_rdf` into `lookup` in https://github.com/metafacture/metafacture-fix/pull/229/commits/c81ae7da85ba8c0a5c6361f62e048db3aec4eb9e. Documentation of `rdf` is not good, I think. Not sure how to do this in an nice way.

Reply:
Conclusion: Only introduce `put_rdfmap()` for now, leave `lookup()` untouched. Investigate later whether to unify `put_*map()` into `put_map()` with map type determined by file extension or explicit `type` parameter (and allow implicit `lookup()` with the same logic) or migrate `put_*map()` (and `lookup()` with named maps) to Catmandu's [store concept](https://librecat.org/Catmandu/#stores).

Reply:
Created issue #279 for the latter part.

Reply:
Special treatment of RDF maps in `lookup()` dropped in 0987eb0.